src/Control/Effect/Cut.hs

```diff
@@ -58,32 +62,90 @@ cut = pure () <|> cutfail
 {-# INLINE cut #-}
 
 
--- | Run a 'Cut' effect within an underlying 'Alternative' instance (typically 'Eff' carrying a 'NonDet' effect).
+-- | Run a 'Cut' effect within an underlying 'Alternative' instance (typically another 'Carrier' for a 'NonDet' effect).
 --
 --   prop> run (runNonDetOnce (runCut (pure a))) == Just a
-runCut :: (Alternative m, Carrier sig m, Effect sig, Monad m) => Eff (CutC m) a -> m a
-runCut = (>>= runBranch (const empty)) . runCutC . interpret
-
-newtype CutC m a = CutC { runCutC :: m (Branch m Bool a) }
-
-instance (Alternative m, Carrier sig m, Effect sig, Monad m) => Carrier (Cut :+: NonDet :+: sig) (CutC m) where
-  ret = CutC . ret . Pure
-  {-# INLINE ret #-}
-
-  eff = CutC . handleSum (handleSum
-    (eff . handle (Pure ()) (bindBranch (ret (None False)) runCutC))
-    (\case
-      Empty    -> ret (None True)
-      Choose k -> runCutC (k True) >>= branch (\ e -> if e then runCutC (k False) else ret (None False)) (\ a -> ret (Alt (ret a) (runCutC (k False) >>= runBranch (const empty)))) (fmap ret . Alt)))
-    (\case
-      Cutfail  -> ret (None False)
-      Call m k -> runCutC m >>= bindBranch (ret (None True)) (runCutC . k))
-    where bindBranch :: (Alternative m, Carrier sig m, Monad m) => m (Branch m Bool a) -> (b -> m (Branch m Bool a)) -> Branch m Bool b -> m (Branch m Bool a)
-          bindBranch cut bind = branch (\ e -> if e then ret (None True) else cut) bind (\ a b -> ret (Alt (a >>= bind >>= runBranch (const empty)) (b >>= bind >>= runBranch (const empty))))
+runCut :: (Alternative m, Carrier sig m) => CutC m a -> m a
+runCut = (>>= runBranch (const empty)) . runCutC' . runCod (CutC' . pure . Pure) . runCutC
+
+newtype CutC m a = CutC { runCutC :: Cod (CutC' m) a }
+  deriving (Applicative, Functor, Monad, MonadFail, MonadIO)
+
+instance (Alternative m, Carrier sig m, Effect sig) => Alternative (CutC m) where
+  empty = send Empty
+  l <|> r = send (Choose (\ c -> if c then l else r))
+
+instance (Alternative m, Carrier sig m, Effect sig) => MonadPlus (CutC m)
+
+instance MonadTrans CutC where
+  lift m = CutC (Cod (\ k -> CutC' (m >>= runCutC' . k)))
+
+instance (Alternative m, Carrier sig m, Effect sig) => Carrier (Cut :+: NonDet :+: sig) (CutC m) where
+  eff = CutC . eff . handleCoercible
+
+
+newtype Cod m a = Cod { unCod :: forall b . (a -> m b) -> m b }
```

Comment:
Can you add some comments here to clear up what’s going on? It’s not immediately apparent to me what `Cod` has to do with implementing `Cut`.

Reply:
I’d like to leave this for #114 since my intention is to remove this hack.